File: src/utils/jsonc.ts
import { readFileSync } from 'node:fs'

/**
 * Strip line (`// ...`) and block (`/* ... *\/`) comments from a JSONC string
 * without mangling comment-like sequences that appear inside string literals
 * (e.g. URLs such as "https://example.com").
 */
export function stripJsonComments(input: string): string {
  let out = ''
  let i = 0
  let inString = false
  while (i < input.length) {
    const c = input[i]
    const next = input[i + 1]

    if (inString) {
      out += c
      if (c === '\\' && next !== undefined) {
        out += next
        i += 2
        continue
      }
      if (c === '"') inString = false
      i++
      continue
    }

    if (c === '"') {
      inString = true
      out += c
      i++
      continue
    }

    if (c === '/' && next === '/') {
      while (i < input.length && input[i] !== '\n') i++
      continue
    }

    if (c === '/' && next === '*') {
      i += 2
      while (i < input.length - 1 && !(input[i] === '*' && input[i + 1] === '/')) i++
      i += 2
      continue
    }

    out += c
    i++
  }
  return out
}

/** Strip comments from JSONC input and parse the result as a JSON object. */
export function parseJsonc(input: string): Record<string, unknown> {
  const stripped = stripJsonComments(input)
  return JSON.parse(stripped) as Record<string, unknown>
}

/** Read a file from disk and parse it as JSONC. */
export function readJsoncFile(path: string): Record<string, unknown> {
  const data = readFileSync(path, 'utf-8')
  return parseJsonc(data)
}

File: packages/cli/internal/jsonc/jsonc.go
// Package jsonc provides JSONC (JSON with comments) support.
// Ported from the TypeScript utilities in src/utils/jsonc.ts.
package jsonc

import (
	"encoding/json"

	"github.com/rluisb/lazyai/packages/cli/internal/files"
)

// StripComments removes single-line (//) and multi-line (/* */) comments
// from a JSONC byte slice without modifying comment-like sequences inside
// string literals (e.g. URLs like "https://example.com").
func StripComments(input []byte) []byte {
	out := make([]byte, 0, len(input))
	i := 0
	inString := false
	n := len(input)

	for i < n {
		c := input[i]
		var next byte
		if i+1 < n {
			next = input[i+1]
		}

		if inString {
			out = append(out, c)
			if c == '\\' && next != 0 {
				out = append(out, next)
				i += 2
				continue
			}
			if c == '"' {
				inString = false
			}
			i++
			continue
		}

		if c == '"' {
			inString = true
			out = append(out, c)
			i++
			continue
		}

		if c == '/' && next == '/' {
			// Skip single-line comment until end of line.
			for i < n && input[i] != '\n' {
				i++
			}
			continue
		}

		if c == '/' && next == '*' {
			// Skip multi-line comment.
			i += 2
			for i < n-1 {
				if input[i] == '*' && input[i+1] == '/' {
					i += 2
					break
				}
				i++
			}
			// If we reached the end without finding closing */, just stop.
			if i >= n-1 {
				i = n
			}
			continue
		}

		out = append(out, c)
		i++
	}

	return out
}

// ParseJSONC strips comments from input and parses the result as JSON into a map.
func ParseJSONC(input []byte) (map[string]any, error) {
	stripped := StripComments(input)
	var result map[string]any
	if err := json.Unmarshal(stripped, &result); err != nil {
		return nil, err
	}
	return result, nil
}

// ReadJSONCFile reads a file and parses it as JSONC.
func ReadJSONCFile(path string) (map[string]any, error) {
	data, err := files.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return ParseJSONC(data)
}
